GPII version updater

mrtyler opened the pull request:
For https://issues.gpii.net/browse/GPII-2542.

Hopefully the README explains what's going on. If it doesn't, please suggest improvements :).

mrtyler committed:
Initial.

These files began their lives at
https://github.com/gpii-ops/gpii-terraform/tree/4a4449de4184d0e1645f14dcd8d1a264071244fd/modules/deploy.

mrtyler committed:
Add CI.

gtirloni commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,13 @@
+build:
+  stage: build
+  script:
+    - docker build -t gpii/version-updater .
```
I'd suggest adding `-rm=true` so intermediate containers don't pile up over time.

mrtyler replied:
Good catch. Looks like it's on by default:

`      --rm                         Remove intermediate containers after a successful build (default true)`

But I added it anyway for explicitness.

gtirloni commented on Dockerfile:
```diff
@@ -0,0 +1,6 @@
+RUN apk update
+RUN apk add git openssh-client curl jq
+COPY update-version-wrapper /
+CMD /update-version-wrapper
```
It'd be nice to have this running under a specific user (see https://github.com/GPII/universal/blob/master/Dockerfile for an example, although here the 'node' user doesn't exist obviously -- "nobody" is a good candidate)

mrtyler replied:
`nobody` doesn't seem like a good choice since it has no login shell and its home dir is /, where it has no permissions to do anything.

I created a dedicated user `app`.

mrtyler replied:
This turned out to be a bit tricky. The new `app` user in the container can't access the volume that is passed in with the ssh creds. I couldn't make it work even with the files owned by the same UID in the container and on the host.

I guess I need to use data volumes or adopt one of the other strategies here: https://stackoverflow.com/questions/23544282/what-is-the-best-way-to-manage-permissions-for-docker-shared-volumes

I'm still working on this.

mrtyler replied:
It works now without any further changes. I think what happened was:
* I must have missed a step getting my local changes over to i46 (forgetting to push to Docker Hub seems the most likely culprit).
* I forgot the selinux `:Z` flag on the volume in my manual testing on the command line (although there is nothing in the audit.log). The ansible code has `:Z` and so the full pipeline was working even.

Anyway, the version of the image that drops root privileges is running successfully on i46.

gtirloni commented on LICENSE:
```diff
@@ -0,0 +1,29 @@
+BSD 3-Clause License
+
+Copyright (c) 2017, Tyler Roscoe
```
I have noticed some repositories in the GPII org have this added to them:

```
All modules are Copyright 2014 Raising the Floor - International except
where noted otherwise in the code itself, or if the modules reside in a
separate directory, they may contain explicit declarations of copyright
in both the LICENSE file in the directory in which they reside and in the
code itself. No external contributions are allowed under licenses which are
fundamentally incompatible with the BSD or Apache licenses that the GPII is
distributed under.
```

mrtyler replied:
Do we have official guidance on Licenses?

If so, point me to it and I'm happy to use it.

If not, I'd rather stick with the generic license until we have a formal decision.

colinbdclark replied:
https://wiki.gpii.net/w/Licensing

The copyright should be attributed to your  organization—in this case, Raising the Floor U.S. If you make substantive changes to a file, you can add your organization to the copyright header.

mrtyler replied:
Thanks Colin. I updated LICENSE (now LICENSE.txt) per those instructions.

gtirloni commented on update-version:
```diff
@@ -0,0 +1,58 @@
+echo "---" >> $outfile
+# Note that '-e' doesn't work with OS X (BSD) echo, which emits a literal '-e'
+# instead :3. Without '-e', GNU echo prints literal '\n' :3 :3.
+echo -e "$yaml" | sort >> $outfile
```
We don't have a policy on this but I usually run shell scripts through shellcheck:

```
Line 29:
        https://index.docker.io/v2/$image/manifests/$tag \
                                   ^-- SC2086: Double quote to prevent globbing and word splitting.
                                                    ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 48:
    yaml="$yaml\n$service: \"$image@$digest\""
               ^-- SC1117: Backslash is literal in "\n". Prefer explicit escaping: "\\n".
 
Line 58:
echo -e "$yaml" | sort >> $outfile
     ^-- SC2039: In POSIX sh, echo flags are undefined.
```

gtirloni commented on update-version-wrapper:
```diff
@@ -0,0 +1,29 @@
+    sleep_secs=300
+    echo "Sleeping $sleep_secs..."
+    sleep $sleep_secs
+done
```
shellcheck warnings:

```
Line 10:
cd gpii-terraform/modules/deploy
^-- SC2164: Use 'cd ... || exit' or 'cd ... || return' in case cd fails.
```

mrtyler replied:
Hmm, my shellcheck (version 0.3.8) doesn't report this one. I switched to `set -e` to guard against unrecoverable errors.

gtirloni commented on README.md:
```diff
@@ -0,0 +1,9 @@
+* `update-version`, which calculates the latest sha256 for each component and writes `version.yml` in the [gpii-terraform repo](https://github.com/gpii-ops/gpii-terraform/).
+* `update-version-wrapper`, a script that runs `update-version` in a loop, committing and pushing `version.yml` if it changes.
+* `Dockerfile`, to build a Docker image that runs `update-version-wrapper`.
+   * A container based on this Docker image is deployed to `i46` and managed by [ansible](https://github.com/inclusive-design/ops).
```
I would be interesting to document what kind of permissions are needed in GitLab/GitHub, if any.

mrtyler replied:
I added some links.

gtirloni commented on update-version:
```diff
@@ -0,0 +1,58 @@
+COMPONENTS="\
+    flow-manager|gpii/universal|latest \
+    preferences-dataloader|gpii/preferences-dataloader|latest \
+    preferences|gpii/universal|latest \
```
If this is loaded from a separate file (say, components.conf), it becomes more obvious what admins are expected to change. Or maybe add a comment here saying this is the only place that needs to be updated to include a new component.

mrtyler replied:
Good idea.

gtirloni commented on update-version:
```diff
@@ -0,0 +1,58 @@
+    api_token=$(\
+        curl -L -s \
+        "https://auth.docker.io/token?service=registry.docker.io&scope=repository:$image:pull" \
+        | jq -r .token \
```
I have used a retry parameters with curl to react quickly to transient issues: https://github.com/GPII/universal/blob/master/scripts/vagrantCloudBasedContainers.sh#L55

I'm not sure how often this is scheduled so suggesting it here in case it makes sense.

mrtyler replied:
Good idea. Stole flags from your link :)

mrtyler committed:
docker build --rm

mrtyler committed:
Run update-version-wrapper under dedicated role user, app.

mrtyler committed:
Extract compoments list to components.conf.

Fix up some path issues exposed by the need to load a new file, including some that were
hidden because I forgot to push the commit that removes these files from gpii-terraform.

mrtyler committed:
Add retries to curl commands.

mrtyler committed:
Fix shellcheck warnings.

Go back to orig_dir strategy for finding update-version from within gpii-terraform sandbox.

mrtyler committed:
Add 'set -e' to bail out during required initial steps.

mrtyler committed:
Update LICENSE per https://wiki.gpii.net/w/Licensing (thanks Colin).

mrtyler committed:
Rename per https://wiki.gpii.net/w/Licensing.

mrtyler committed:
Add some links.

gpii-bot committed:
Clean out sandbox, as it is left behind after container restart.

mrtyler committed:
Add a timestamp at the end of the cycle.

mrtyler commented:
This PR says "merged" because I accidentally pushed to the gpii-ops repo (instead of gitlab), but it is still awaiting review.

gtirloni commented:
LGTM

mrtyler commented:
Merged manually since Github thinks this was already merged.
